common/functions/src/aggregates/aggregate_retention.rs

```diff
@@ -0,0 +1,224 @@
+    }
+
+    fn deserialize(&mut self, reader: &mut &[u8]) -> Result<()> {
+        self.events = deserialize_from_slice(reader)?;
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct AggregateRetentionFunction {
+    display_name: String,
+    events_size: u8,
+    _arguments: Vec<DataField>,
+}
+
+impl AggregateFunction for AggregateRetentionFunction {
+    fn name(&self) -> &str {
+        "AggregateRetentionFunction"
+    }
+
+    fn return_type(&self) -> Result<DataTypePtr> {
+        Ok(JsonValue::to_data_type())
+    }
+
+    fn init_state(&self, place: StateAddr) {
+        place.write(|| AggregateRetentionState { events: 0 });
+    }
+
+    fn state_layout(&self) -> std::alloc::Layout {
+        Layout::new::<AggregateRetentionState>()
+    }
+
+    fn accumulate(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        _validity: Option<&common_arrow::arrow::bitmap::Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let new_columns: Vec<&BooleanColumn> = columns
+            .iter()
+            .map(|column| Series::check_get(column).unwrap())
+            .collect();
+        for i in 0..input_rows {
+            for j in 0..self.events_size {
+                if new_columns[j as usize].get_data(i) {
+                    state.add(j);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn accumulate_row(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        row: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let new_columns: Vec<&BooleanColumn> = columns
+            .iter()
+            .map(|column| Series::check_get(column).unwrap())
+            .collect();
+        for j in 0..self.events_size {
+            if new_columns[j as usize].get_data(row) {
+                state.add(j);
+            }
+        }
+        Ok(())
+    }
+
+    fn serialize(&self, place: StateAddr, writer: &mut BytesMut) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.serialize(writer)
+    }
+
+    fn deserialize(&self, place: StateAddr, reader: &mut &[u8]) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.deserialize(reader)
+    }
+
+    fn merge(&self, place: StateAddr, rhs: StateAddr) -> Result<()> {
+        let rhs = rhs.get::<AggregateRetentionState>();
+        let state = place.get::<AggregateRetentionState>();
+        state.merge(rhs);
+        Ok(())
+    }
+
+    #[allow(unused_mut)]
+    fn merge_result(
+        &self,
+        place: StateAddr,
+        array: &mut dyn common_datavalues::MutableColumn,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let builder: &mut MutableObjectColumn<JsonValue> = Series::check_get_mutable_column(array)?;
```

Comment:
```suggestion
        let builder = Series::check_get_mutable_column(array)?;
```

How about removing type annotations unless rust can't determine them?

This will make life a bit easier :smile:

Reply:
It can not compiled if we ignore it. It needs a type here. compiler says:

```
type annotations needed for `&mut T`
type must be known at this pointrustc[E0282](https://doc.rust-lang.org/error-index.html#E0282)
[aggregate_retention.rs(147, 13): ]()consider giving `builder` the explicit type `&mut T`, where the type parameter `T` is specified
```